Backend/Remora.Discord.Caching/CacheKey.cs

```diff
@@ -0,0 +1,179 @@
+//
+//  CacheKey.cs
+//
+//  Author:
+//       Jarl Gullberg <jarl.gullberg@gmail.com>
+//
+//  Copyright (c) 2017 Jarl Gullberg
+//
+//  This program is free software: you can redistribute it and/or modify
+//  it under the terms of the GNU Lesser General Public License as published by
+//  the Free Software Foundation, either version 3 of the License, or
+//  (at your option) any later version.
+//
+//  This program is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//  GNU Lesser General Public License for more details.
+//
+//  You should have received a copy of the GNU Lesser General Public License
+//  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+//
+
+using System;
+using System.Diagnostics;
+
+namespace Remora.Discord.Caching
+{
+    /// <summary>
+    /// Represents a cache key.
+    /// </summary>
+    [DebuggerDisplay("{" + nameof(Key) + ",nq}")]
+    public readonly struct CacheKey : IEquatable<CacheKey>
```

Comment:
This structure has a relatively heavy maintenance impact, considering that the argument count may grow in the future. See if you can generalize it by using a list or preallocated array (a `params` constructor, perhaps?).

Reply:
I can change this to an array, however the only downside is that every time a cache key is being constructed, a new array will be allocated. Next to that, you'll have to loop through the array to make a hash code.

Reply:
Thinking outside the box for a moment, is there a better way of creating a cache key?

Reply:
I've added a new constructor that accepts an object array. When using less than 3 keys it doesn't create an object array and it uses the arg1..3 field (https://github.com/Nihlus/Remora.Discord/pull/58/commits/a514688a0f8d2e4c8c937e145b00d8be0b0296d0).

> Thinking outside the box for a moment, is there a better way of creating a cache key?

We could also just use a string in the `MemoryCache`-implemetation, instead of relying on the hash code.

Reply:
I'd prefer not to have a hybrid solution; that's unneeded complexity. Choose one or the other.

String allocations should be avoided where possible, so relying on the hash code is preferable in that case.

Reply:
Using a string as key in Redis is very useful to look up the values:

![image](https://user-images.githubusercontent.com/2109929/116119556-db68af00-a6be-11eb-942f-3f0d87656bd3.png)

Also, this way you can easily remove cache (for instance, removing all values with the pattern `IGuild:*`).
So honestly I would like to have a string as key in Redis.

Reply:
For Redis, that's totally fine; I was more referring to the memory cache that currently uses the hashcode (and I think should continue to do so).

Reply:
It does... That's why I made the CacheKey. So I'm unsure what has to be done here now.

Reply:
It's two things.

  1) Either remove the arg1, arg2, ... fields and just use the array, or remove the array and just use the arg1, arg2 fields. I would prefer that you did the former.
  2) Don't do anything to the keying - redis using strings and memory cache using the struct directly is perfectly fine.